src/qml/qmlwaveformdisplay.cpp

```diff
@@ -245,7 +246,55 @@ void QmlWaveformDisplay::slotWaveformUpdated() {
 }
 
 QQmlListProperty<QmlWaveformRendererFactory> QmlWaveformDisplay::renderers() {
-    return {this, &m_waveformRenderers};
+    return {this,
+            nullptr,
+            &QmlWaveformDisplay::renderers_append,
+            &QmlWaveformDisplay::renderers_count,
+            &QmlWaveformDisplay::renderers_at,
+            &QmlWaveformDisplay::renderers_clear};
+}
+
+// Static
+void QmlWaveformDisplay::renderers_append(
+        QQmlListProperty<QmlWaveformRendererFactory>* pList,
+        QmlWaveformRendererFactory* value) {
+    QmlWaveformDisplay* pWaveform = static_cast<QmlWaveformDisplay*>(pList->object);
+    VERIFY_OR_DEBUG_ASSERT(pWaveform) {
+        return;
+    }
+    pWaveform->m_dirtyFlag.setFlag(DirtyFlag::Window, true);
+    pWaveform->m_waveformRenderers.append(value);
+}
+
+// Static
+qsizetype QmlWaveformDisplay::renderers_count(QQmlListProperty<QmlWaveformRendererFactory>* pList) {
+    QmlWaveformDisplay* pWaveform = static_cast<QmlWaveformDisplay*>(pList->object);
+    VERIFY_OR_DEBUG_ASSERT(pWaveform) {
+        return 0;
+    }
+    pWaveform->m_dirtyFlag.setFlag(DirtyFlag::Window, true);
+    return pWaveform->m_waveformRenderers.count();
+}
+
+// Static
+QmlWaveformRendererFactory* QmlWaveformDisplay::renderers_at(
+        QQmlListProperty<QmlWaveformRendererFactory>* pList, qsizetype index) {
+    QmlWaveformDisplay* pWaveform = static_cast<QmlWaveformDisplay*>(pList->object);
+    VERIFY_OR_DEBUG_ASSERT(pWaveform) {
+        return nullptr;
+    }
```

Comment:
is this supposed to check if the downcast was successful or just if `pList->object` is not nullptr?

Reply:
Just if pList->object is not nullptr - this is very unlikely as we create the list with a parent. Would you prefer this to be before the var def?

Reply:
yeah, lets do it before the cast. otherwise it looks like someone meant to use a `dynamic_cast` but used a `static_cast` by accident.